### PR TITLE
feat: onboarding mode and edit method

### PR DESCRIPTION
**Pull request by almeidx**
**Please describe the changes this PR makes and why it should be merged:**
Split this up from #9120 so we can have some onboarding stuff to work with

Upstream:
- discord/discord-api-docs#6101

Depends on:
- https://github.com/discordjs/discord-api-types/pull/773
- #9880

**Status and versioning classification:**

### CODE REVIEW

**Comment by Jiralite** [comment]
This needs a rebase.

This could also be split up again—https://github.com/discord/discord-api-docs/pull/6041 only has minor changes and would no longer block this.

File: packages/discord.js/src/structures/Guild.js [hunk 1]
```diff
@@ -861,6 +871,70 @@ class Guild extends AnonymousGuild {
...
   }
 
+  /**
+   * Options used to edit the guild onboarding.
+   * @typedef {Object} GuildOnboardingEditOptions
+   * @property {GuildOnboardingPromptData[]} prompts The prompts shown during onboarding
+   * and in customize community
+   * @property {ChannelResolvable[]} defaultChannels The channels that new members get opted into automatically
+   * @property {boolean} enabled Whether the onboarding is enabled
+   * @property {GuildOnboardingMode} mode The mode to edit the guild onboarding with
```
[the remark this answers is not in the document]
Comment (Jiralite) [reply packages/discord.js/src/structures/Guild.js:881]: They appear not to be documented as such though.

**Comment by Jiralite** [comment]
Moved this into a draft state due to strange documented behaviour that has been brought up internally.[^1] This should be resolve soon.

[^1]: https://discord.com/channels/222078108977594368/999796393810067607/1139253705078734948

File: packages/discord.js/typings/index.d.ts [hunk 2]
```diff
@@ -5701,6 +5704,31 @@ export type GuildTemplateResolvable = string;
 
 export type GuildVoiceChannelResolvable = VoiceBasedChannel | Snowflake;
 
+export interface GuildOnboardingEditOptions {
+  prompts: GuildOnboardingPromptData[];
```
Comment (Jiralite) [inline packages/discord.js/typings/index.d.ts:5708]:
```suggestion
  prompts: readonly GuildOnboardingPromptData[];
```

File: packages/discord.js/typings/index.d.ts [hunk 3]
```diff
@@ -5701,6 +5704,31 @@ export type GuildTemplateResolvable = string;
...
+  defaultChannels: ChannelResolvable[];
```
Comment (Jiralite) [inline packages/discord.js/typings/index.d.ts:5709]:
```suggestion
  defaultChannels: readonly ChannelResolvable[];
```

File: packages/discord.js/typings/index.d.ts [hunk 4]
```diff
@@ -5701,6 +5704,31 @@ export type GuildTemplateResolvable = string;
...
+  reason?: string;
+}
+
+export interface GuildOnboardingPromptData {
+  title: string;
+  singleSelect: boolean;
+  required: boolean;
+  inOnboarding: boolean;
+  type: GuildOnboardingMode;
+  options: GuildOnboardingPromptOptionData[];
```
Comment (Jiralite) [inline packages/discord.js/typings/index.d.ts:5721]:
```suggestion
  options: readonly GuildOnboardingPromptOptionData[];
```

File: packages/discord.js/typings/index.d.ts [hunk 5]
```diff
@@ -5701,6 +5704,31 @@ export type GuildTemplateResolvable = string;
...
+}
+
+export interface GuildOnboardingPromptOptionData {
+  channels: ChannelResolvable[];
```
Comment (Jiralite) [inline packages/discord.js/typings/index.d.ts:5725]:
```suggestion
  channels: readonly ChannelResolvable[];
```

File: packages/discord.js/typings/index.d.ts [hunk 6]
```diff
@@ -5701,6 +5704,31 @@ export type GuildTemplateResolvable = string;
...
+  roles: RoleResolvable[];
```
Comment (Jiralite) [inline packages/discord.js/typings/index.d.ts:5726]:
```suggestion
  roles: readonly RoleResolvable[];
```

**Review by vladfrangu — APPROVED** [review]

**Comment by Jiralite** [comment]
This needs a rebase.

File: packages/discord.js/src/structures/Guild.js [hunk 7]
```diff
@@ -871,6 +872,72 @@ class Guild extends AnonymousGuild {
...
+   * @property {ChannelResolvable[]} [defaultChannels] The channels that new members get opted into automatically
+   * @property {boolean} [enabled] Whether the onboarding is enabled
+   * @property {GuildOnboardingMode} [mode] The mode to edit the guild onboarding with
+   * @property {string} [reason] The reason for editing the guild onboarding
+   */
+
+  /**
+   * Data for editing a guild onboarding prompt.
+   * @typedef {Object} GuildOnboardingPromptData
+   * @property {string} [id] The id of the prompt
```
Comment (Jiralite) [inline packages/discord.js/src/structures/Guild.js:888]: Shouldn't this be a `Snowflake`?

File: packages/discord.js/src/structures/Guild.js [hunk 8]
```diff
@@ -871,6 +872,72 @@ class Guild extends AnonymousGuild {
...
+          type: prompt.type,
+          options: prompt.options.map(option => ({
+            channel_ids: option.channels.map(c => this.channels.resolveId(c)),
+            role_ids: option.roles.map(r => this.roles.resolveId(r)),
+            title: option.title,
+            description: option.description,
+            emoji: this.emojis.resolveIdentifier(option.emoji),
+          })),
+        })),
+        default_channel_ids: options.defaultChannels?.map(c => this.channels.resolveId(c)),
```
Comment (Jiralite) [inline packages/discord.js/src/structures/Guild.js:931]: I'm not too sure I'm a fan of the single-charactered variable.

Comment (Jiralite) [inline packages/discord.js/src/structures/Guild.js:931]: The channel ids may take a collection of channels, which is what's returned from `fetchOnboarding()`. Would it be possible to add this as a possible type (the collection) to `editOnboarding()`?

Comment (Jiralite) [inline packages/discord.js/src/structures/Guild.js:931]: Semantically speaking, shouldn't these be `this.guild.channels`?

File: packages/discord.js/src/structures/Guild.js [hunk 9]
```diff
@@ -871,6 +872,72 @@ class Guild extends AnonymousGuild {
...
+
+  /**
+   * Edits the guild onboarding data for this guild.
+   * @param {GuildOnboardingEditOptions} options The options to provide
+   * @returns {Promise<GuildOnboarding>}
+   */
+  async editOnboarding(options) {
+    const newData = await this.client.rest.put(Routes.guildOnboarding(this.id), {
+      body: {
+        prompts: options.prompts?.map(prompt => ({
```
Comment (Jiralite) [inline packages/discord.js/src/structures/Guild.js:915]: The prompts may take a collection of prompts, which is what's returned from `fetchOnboarding()`. Would it be possible to add this as a possible type (the collection) to `editOnboarding()`?

File: packages/discord.js/src/structures/Guild.js (hunk 8 above)
Comment (almeidx) [reply packages/discord.js/src/structures/Guild.js:941]: This method is in the `Guild` class, which doesn't have a `guild` property

File: packages/discord.js/src/structures/Guild.js [hunk 10]
```diff
@@ -881,6 +882,83 @@ class Guild extends AnonymousGuild {
...
+  /**
+   * Data for editing a guild onboarding prompt option.
+   * @typedef {Object} GuildOnboardingPromptOptionData
+   * @property {ChannelResolvable[]|Collection<Snowflake, GuildChannel>} [channels]
+   * The channels a member is added to when the option is selected
+   * @property {RoleResolvable[]|Collection<Snowflake, Role>} [roles]
+   * The roles assigned to a member when the option is selected
+   * @property {string} title The title of the option
+   * @property {?string} [description] The description of the option
+   * @property {?(EmojiIdentifierResolvable|GuildOnboardingPromptOptionEmoji)} [emoji] The emoji of the option
```
Comment (Jiralite) [inline packages/discord.js/src/structures/Guild.js:919]: `GuildOnboardingPromptOptionEmoji` is superfluous to add here.

File: packages/discord.js/typings/index.d.ts [hunk 11]
```diff
@@ -5774,6 +5777,32 @@ export type GuildTemplateResolvable = string;
...
+  type?: GuildOnboardingPromptType;
+  options: readonly GuildOnboardingPromptOptionData[] | ReadonlyCollection<Snowflake, GuildOnboardingPromptOption>;
+}
+
+export interface GuildOnboardingPromptOptionData {
+  channels?: readonly ChannelResolvable[] | ReadonlyCollection<Snowflake, GuildChannel>;
+  roles?: readonly RoleResolvable[] | ReadonlyCollection<Snowflake, Role>;
+  title: string;
+  description?: string | null;
+  emoji?: EmojiIdentifierResolvable | GuildOnboardingPromptOptionEmoji | null;
```
Comment (Jiralite) [inline packages/discord.js/typings/index.d.ts:5803]: `GuildOnboardingPromptOptionEmoji` is superfluous to add here.

**Review by Jiralite — CHANGES_REQUESTED** [review]

**Review by Jiralite — APPROVED** [review]

File: packages/discord.js/src/structures/GuildOnboardingPromptOption.js [hunk 12]
```diff
@@ -79,6 +72,15 @@ class GuildOnboardingPromptOption extends Base {
   get guild() {
     return this.client.guilds.cache.get(this.guildId);
   }
+
+  /**
+   * The emoji of this onboarding prompt option
+   * @type {GuildEmoji|Emoji}
```
Comment (vladfrangu) [inline packages/discord.js/src/structures/GuildOnboardingPromptOption.js:78]: This can be null

**Review by vladfrangu — DISMISSED** [review]

**Review by vladfrangu — APPROVED** [review]

**Review by suneettipirneni — APPROVED** [review]

**Pull request merged**